text_extensions_for_pandas/array/tensor.py

```diff
@@ -0,0 +1,223 @@
+#
+#  Copyright (c) 2020 IBM Corp.
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+#
+# tensor.py
+#
+# Part of text_extensions_for_pandas
+#
+# Pandas extensions to support columns of N-dimensional tensors of equal shape.
+#
+
+from typing import *
+
+import numpy as np
+import pandas as pd
+from memoized_property import memoized_property
+
+# Internal imports
+import text_extensions_for_pandas.util as util
+
+
+@pd.api.extensions.register_extension_dtype
+class TensorType(pd.api.extensions.ExtensionDtype):
+    """
+    Panda datatype for a span that represents a range of characters within a
+    target string.
+    """
+
+    @property
+    def type(self):
+        # The type for a single row of a column of type CharSpan
+        return np.ndarray
+
+    @property
+    def name(self) -> str:
+        """A string representation of the dtype."""
+        return "Tensor"
+
+    @classmethod
+    def construct_from_string(cls, string: str):
+        """
+        See docstring in `ExtensionDType` class in `pandas/core/dtypes/base.py`
+        for information about this method.
+        """
+        # Upstream code uses exceptions as part of its normal control flow and
+        # will pass this method bogus class names.
+        if string == cls.__name__:
+            return cls()
+        else:
+            raise TypeError(
+                f"Cannot construct a '{cls.__name__}' from '{string}'")
+
+    @classmethod
+    def construct_array_type(cls):
+        """
+        See docstring in `ExtensionDType` class in `pandas/core/dtypes/base.py`
+        for information about this method.
+        """
+        return TensorArray
+
+
+class TensorArray(pd.api.extensions.ExtensionArray):
+    """
+    A Pandas `ExtensionArray` that represents a column of `numpy.ndarray`s,
+    or tensors, where the outer dimension is the count of tensors in the column.
+    """
+
+    def __init__(self, values: Union[np.ndarray, List[np.ndarray]],
```

Comment:
I've started using `Sequence` instead of `List` in public APIs' type hints. I see that you found a place where my code was still using `List` and copied that :-). I'll fix the constructors for `TokenSpan` and `CharSpan` in a separate commit. Can you change the `List[np.ndarray]` to `Sequence[np.ndarray]` here?

Reply:
Sure, good idea